File: MergeFile.py
import re
from os import walk

outputFile = open("ACGNs-SoftwareScript.user.js", "w", encoding = 'utf8')
outputFile.write("/* 本檔由 MergeFile.py 自動產生, 欲修改code請至src資料夾 */" + "\n")

def findFileName(line):
    a = re.search("[A-Za-z0-9_]*[\"\'];$", line).group()
    fileName = re.search("[A-Za-z0-9_]*", a).group()
    return fileName


fileNameList = []
fileRootList = []
importList = []
writeCount = -1

def writeIntoFile(line):
    if (re.match("export ", line)):
        line = line.lstrip("export ")
    outputFile.write(line)

requireList = []
def copyJs(jsFile):
    global writeCount
    writeCount += 1
    dash = ""
    for i in range(writeCount):
        dash += "-"
    outputFile.write("//" + dash + "start file: " + jsFile + "\n")

    readJsFile(jsFile)

    outputFile.write("//" + dash + "end file: " + jsFile + "\n")
    space = ""
    for i in range(writeCount):
        space += " "
    outputFile.write("//" + space + "===========================\n")
    writeCount -= 1


def readJsFile(jsFile):
    with open(jsFile, "r", encoding = 'utf8') as js:
        for line in js:
            print(line, end="")
            if (re.match("import", line)):
                importFile(line)
            elif (re.search("require", line)):
                requireFile(line)
            elif (re.search("eslint-enable", line)):
                print("//skip eslint-enable")
            else:
                writeIntoFile(line)


def importFile(line):
    fileName = findFileName(line)
    if (importList.count(fileName) < 1):
        print("//new import")
        importList.append(fileName)
        importRoot = fileRootList[fileNameList.index(fileName + ".js")]
        copyJs(importRoot)
    else:
        print("//import already exists")


def requireFile(line):
    if (requireList.count(line) < 1):
        print("//new require")
        requireList.append(line)
        writeIntoFile(line)
    else:
        print("//require already exists")


  
for root, dirs, files in walk("./src"):
    # print("路徑：", root)
    # print("  目錄：", dirs)
    # print("  檔案：", files)
    for f in files:
        fileRoot = root + "/" + f
        print(fileRoot)
        fileNameList.append(f)
        fileRootList.append(fileRoot)
mainRoot = fileRootList[fileNameList.index("main.js")]
copyJs(mainRoot)

outputFile.close()

File: mergeFile.js
const fs = require('fs');
const getVersion = require('./version.js');

const srcDirPath = './src/';
const mainFilePath = 'main'; // 從 srcDirPath 底下開始尋找
const outputFilePath = './ACGNs-SoftwareScript.user.js';

const debugMode = false;
function debugConsole(...messages) {
  if (debugMode) {
    console.log(...messages);
  }
}


let writeCount = -1;
function copyJsFile(filePath) {
  writeCount += 1;
  let dashLine = '';
  for (let i = 0; i < writeCount; i += 1) {
    dashLine += '-';
  }
  fs.writeSync(outputFile, `//${dashLine}start file: ${filePath}\n`);
  analyzeFileAndOutput(filePath);
  fs.writeSync(outputFile, `//${dashLine}end file: ${filePath}\n`);
  fs.writeSync(outputFile, `//\n`);
  writeCount -= 1;
}

function analyzeFileAndOutput(filePath) {
  const lines = fs.readFileSync(srcDirPath + filePath + '.js', 'utf8').split('\n');
  for (const line of lines) {
    debugConsole(line);
    if (isStartWith('import ', line)) {
      importFile(line);
    }
    else if (isStartWith('// @version', line)) {
      fs.writeSync(outputFile, `// @version      ${getVersion()}\n`);
    }
    else if (isStartWith('export ', line)) {
      fs.writeSync(outputFile, line.slice(7) + '\n');
    }
    else {
      fs.writeSync(outputFile, line + '\n');
    }
  }
}

function isStartWith(regexp, line) {
  return line.search(regexp) === 0;
}

const importList = [];
function importFile(line) {
  const filePath = findImportPath(line);
  const isImported = importList.find((imported) => {
    return imported === filePath;
  });

  if (! isImported) {
    debugConsole(`//new import: ${filePath}`);
    importList.push(filePath);
    copyJsFile(filePath);
  }
  else {
    debugConsole(`//import already exists`);
  }
}

function findImportPath(line) {
  const start = line.search(/from '/) + 6;
  const end = line.search(/';/);
  const path = line.slice(start, end);

  return path;
}


const outputFile = fs.openSync(outputFilePath, 'w');
fs.writeSync(outputFile, '/* 本檔由 mergeFile.js 自動產生, 欲修改code請至src資料夾 */\n');
copyJsFile(mainFilePath);
fs.closeSync(outputFile);

console.log('');
console.log('merge file done.');
